src/Aspire.Hosting.Azure.Storage/AzureStorageExtensions.cs

```diff
@@ -326,10 +275,41 @@ public static IResourceBuilder<AzureBlobStorageContainerResource> AddBlobContain
         blobContainerName ??= name;
 
         AzureBlobStorageContainerResource resource = new(name, blobContainerName, builder.Resource);
-
         builder.Resource.Parent.BlobContainers.Add(resource);
 
-        return builder.ApplicationBuilder.AddResource(resource);
+        string? connectionString = null;
+        builder.ApplicationBuilder.Eventing.Subscribe<ConnectionStringAvailableEvent>(resource, async (@event, ct) =>
+        {
+            var parentResource = resource.Parent;
+            connectionString = await parentResource.ConnectionStringExpression.GetValueAsync(ct).ConfigureAwait(false) ?? throw new DistributedApplicationException($"{nameof(ConnectionStringAvailableEvent)} was published for the '{parentResource.Name}' resource but the connection string was null.");
+        });
+
+        var healthCheckKey = $"{resource.Name}_check";
+
+        BlobContainerClient? blobContainerClient = null;
+        var healthCheckRegistration = new HealthCheckRegistration(
+            healthCheckKey,
+            sp =>
+            {
+                if (blobContainerClient is null)
+                {
+                    var blobServiceClient = CreateBlobServiceClient(connectionString ?? throw new InvalidOperationException("Connection string client is not initialized."));
+                    blobContainerClient = blobServiceClient.GetBlobContainerClient(blobContainerName);
+                    blobContainerClient.CreateIfNotExists();
```

Comment:
I still don't like mutating state during a health check, even if it is the first time. I don't get why we can't follow the same pattern used in CosmosDB and Sql/Postgres.

Reply:
My understanding is that we can create the dbs when CosmosDB or Postgres or SqlServer is ready. We can connect to these and have a health check for these so on the ready event we know we can create the dbs.

For storage there is nothing to connect to, no connection string, there isn't a connection string for it, the blob is the thing.

Maybe the reason why there is no health check for storage itself either.